toolchain/check/merge.cpp

```diff
@@ -232,78 +229,101 @@ static auto CheckRedeclParam(Context& context, bool is_implicit_param,
         .Emit();
   };
 
-  auto new_param_pattern = context.insts().Get(new_param_pattern_id);
-  auto prev_param_pattern = context.insts().Get(prev_param_pattern_id);
-  if (new_param_pattern.kind() != prev_param_pattern.kind()) {
-    emit_diagnostic();
-    return false;
-  }
+  struct PatternPair {
+    SemIR::InstId prev_id;
+    SemIR::InstId new_id;
+  };
 
-  if (new_param_pattern.Is<SemIR::AddrPattern>()) {
-    new_param_pattern_id = new_param_pattern.As<SemIR::AddrPattern>().inner_id;
-    new_param_pattern = context.insts().Get(new_param_pattern_id);
-    prev_param_pattern_id =
-        prev_param_pattern.As<SemIR::AddrPattern>().inner_id;
-    prev_param_pattern = context.insts().Get(prev_param_pattern_id);
-    if (new_param_pattern.kind() != prev_param_pattern.kind()) {
-      emit_diagnostic();
-      return false;
-    }
-  }
+  llvm::SmallVector<PatternPair, 1> pattern_stack;
 
-  if (new_param_pattern.Is<SemIR::AnyParamPattern>()) {
-    new_param_pattern_id =
-        new_param_pattern.As<SemIR::ValueParamPattern>().subpattern_id;
-    new_param_pattern = context.insts().Get(new_param_pattern_id);
-    prev_param_pattern_id =
-        prev_param_pattern.As<SemIR::ValueParamPattern>().subpattern_id;
-    prev_param_pattern = context.insts().Get(prev_param_pattern_id);
+  pattern_stack.push_back({.prev_id = orig_prev_param_pattern_id,
+                           .new_id = orig_new_param_pattern_id});
+
+  do {
```

Comment:
What was the motivation to switch to a stack and a loop? Was the code wrong before in some way?